Issue 3599 - Always Scan DDN Params

kingthorin opened the pull request:
* Add VariantDdnPath.
* Modify AbstractAppParamPlugin to handle/leverage VariantDdnPath.
* Modify Messages.properties to combine the DDN scanning option with one
of the default options (URL Query String).

Core help update will be submitted after this is accepted.

thc202 commented on src/org/parosproxy/paros/core/scanner/VariantDdnPath.java:
```diff
@@ -0,0 +1,109 @@
+	public void setMessage(HttpMessage msg) {
+		try {
+			List<String> treePaths = Model.getSingleton().getSession().getTreePath(msg);
+			int i = 0;
```
Could start in 1 instead of adding one when creating the `NameValuePair`, in any case a comment explaining the reason why adding one (or starting at 1) would be better.

kingthorin replied:
Hahaha excellent point, I totally hadn't thought about it from anyone else's perspective, just made it work.

Will tweak.

thc202 commented on src/org/parosproxy/paros/core/scanner/VariantDdnPath.java:
```diff
@@ -0,0 +1,109 @@
+			boolean escaped) {
+		try {
+			URI uri = msg.getRequestHeader().getURI();
+			String[] paths = msg.getRequestHeader().getURI().getPath().toString().split("/");
```
- Can use the variable `uri`;
 - Should use `getEscapedPath()`;
 - `toString()` can be removed (it's already returning a String).

thc202 commented on src/org/parosproxy/paros/core/scanner/VariantDdnPath.java:
```diff
@@ -0,0 +1,109 @@
+			if (originalPair.getPosition() < paths.length) {
+				String encodedValue = (escaped) ? value : getEscapedValue(value);
+				paths[originalPair.getPosition()] = encodedValue;
+				String path = StringUtils.join(paths, "/");
```
Could use `String.join`.

thc202 commented on src/org/parosproxy/paros/core/scanner/VariantDdnPath.java:
```diff
@@ -0,0 +1,109 @@
+				try {
+					uri.setEscapedPath(path);
+				} catch (URIException e) {
+					// Looks like it wasn't escaped after all
```
Worth logging as warn or even error, if this happens it's because of a scanner misusing the variant (i.e. calling `setEscapedParameter` with a non-escaped value).

thc202 commented on src/org/parosproxy/paros/core/scanner/VariantDdnPath.java:
```diff
@@ -0,0 +1,109 @@
+					uri.setPath(path);
+				}
+			}
+		} catch (URIException e) {
```
This catch can be moved to above catch block (no longer needed outside per use of `getEscapedPath` also the level can be changed to debug, not much that can be done if the path is already wrong).

kingthorin commented:
Fixed-up.

I also named the position variable more specifically and added a class lead-in JavaDoc block that explains the variant and position numbering.

thc202 commented:
Thank you!

psiinon commented:
Thanks :) Worth updating the help too? (unless you've already submitted a PR for that;)
